src/shims/env.rs

```diff
@@ -54,33 +63,109 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+                    // If lpBuffer is not large enough to hold the data, the return value is the buffer size, in characters,
+                    // required to hold the string and its terminating null character and the contents of lpBuffer are undefined.
+                    var_size + 1
+                } else {
+                    let buf_ptr = this.read_scalar(buf_op)?.not_undef()?;
+                    for i in 0..var_size {
+                        this.memory.copy(
+                            this.force_ptr(var_ptr.ptr_offset(Size::from_bytes(i) * 2, this)?)?,
+                            this.force_ptr(buf_ptr.ptr_offset(Size::from_bytes(i) * 2, this)?)?,
+                            Size::from_bytes(2),
+                            true,
+                        )?;
+                    }
+                    // If the function succeeds, the return value is the number of characters stored in the buffer pointed to by lpBuffer,
+                    // not including the terminating null character.
+                    var_size
+                };
+                return_val
+            }
+            None => {
+                this.set_last_error(Scalar::from_u32(203))?; // ERROR_ENVVAR_NOT_FOUND
+                0 // return zero upon failure
+            }
+        })
+    }
+
+    fn getenvironmentstringsw(&mut self) -> InterpResult<'tcx, Scalar<Tag>> {
+        let this = self.eval_context_mut();
+        this.assert_target_os("windows", "GetEnvironmentStringsW");
+
+        // Info on layout of environment blocks in Windows: 
+        // https://docs.microsoft.com/en-us/windows/win32/procthread/environment-variables
+        let mut env_vars = std::ffi::OsString::new();
+        for &item in this.machine.env_vars.values()? {
+            let env_var = this.read_os_str_from_target_str(Scalar::from(item))?;
+            env_vars.push(env_var);
+            env_vars.push("\0");
+        }
+
+        // Allocate environment block
+        let tcx = this.tcx;
+        let env_block_size = env_vars.len().checked_add(1).unwrap();
+        let env_block_type = tcx.mk_array(tcx.types.u16, u64::try_from(env_block_size).unwrap());
+        let env_block_place = this.allocate(this.layout_of(env_block_type)?, MiriMemoryKind::WinHeap.into());
+        
+        // Store environment variables to environment block
+        // Final null terminator(block terminator) is pushed by `write_os_str_to_wide_str`
+        this.write_os_str_to_wide_str(&env_vars, env_block_place, u64::try_from(env_block_size).unwrap())?;
+
+        Ok(env_block_place.ptr)
+    }
+
     fn setenv(
         &mut self,
         name_op: OpTy<'tcx, Tag>,
         value_op: OpTy<'tcx, Tag>,
     ) -> InterpResult<'tcx, i32> {
         let mut this = self.eval_context_mut();
+        let target_os = this.tcx.sess.target.target.target_os.as_str();
+        assert!(target_os == "linux" || target_os == "macos", "`setenv` is only available for the UNIX target family");
```

Comment:
So you decided against reusing this for Windows, after all? Then you should switch back to `c_str` (instead of `target_str`) methods in here.